[GStreamer] imported/w3c/web-platform-tests/webcodecs/audio-data.any.html is a permanent failure

philn opened the pull request:
#### aee63a64ceb8f0f02f29e68f53055f14a931d4c4
<pre>
[GStreamer] imported/w3c/web-platform-tests/webcodecs/audio-data.any.html is a permanent failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=284626">https://bugs.webkit.org/show_bug.cgi?id=284626</a>

Reviewed by Xabier Rodriguez-Calvar.

The GStreamer AudioData::copyTo() is now spec compliant, handling the various layout conversion
cases (from/to planar/interleaved) and format changes.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-data.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-data.any.worker-expected.txt: Removed.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
(WebCore::computeCopyElementCount):
* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
(WebCore::GStreamerInternalAudioEncoder::encode):
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
(WebCore::PlatformRawAudioData::create):
(WebCore::PlatformRawAudioDataGStreamer::numberOfFrames const):
(WebCore::PlatformRawAudioDataGStreamer::isInterleaved const):
(WebCore::PlatformRawAudioDataGStreamer::planesOfSamples):
(WebCore::layoutToString):
(WebCore::PlatformRawAudioData::copyTo):
(WebCore::getAudioConvertedForFormat): Deleted.
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::GstMappedAudioBuffer::GstMappedAudioBuffer):
(WebCore::GstMappedAudioBuffer::samples const):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:

Canonical link: <a href="https://commits.webkit.org/290147@main">https://commits.webkit.org/290147@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/536f8c3266d1951dfc2ceaa38db770a49d7b0477

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89034 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8558 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43891 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94008 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39794 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91085 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16743 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68601 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26275 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6836 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80833 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48964 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35213 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38902 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76947 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36194 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95845 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16214 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11856 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77480 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16470 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76624 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76770 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18940 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21160 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19808 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9297 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16228 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21539 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15969 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19420 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17750 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
```diff
@@ -113,7 +113,7 @@ ExceptionOr<size_t> computeCopyElementCount(const WebCodecsAudioData& data, cons
-#if !USE(AVFOUNDATION)
+#if !USE(AVFOUNDATION) && !USE(GSTREAMER)
     // 5. If [[format]] does not equal destFormat and the User Agent does not support the requested AudioSampleFormat conversion, throw a NotSupportedError DOMException. Conversion to f32-planar must always be supported.
     if (options.format && *options.format != destFormat && destFormat != AudioSampleFormat::F32Planar)
```
This can probably be removed now, is there any other supported configuration other than GSTREAMER and AVFOUNDATION?

philn replied:
I don't think so, will remove this ifdef.

jyavenard commented on Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
```diff
@@ -188,67 +174,119 @@ int64_t PlatformRawAudioDataGStreamer::timestamp() const
-    [[maybe_unused]] auto [sourceFormat, sourceLayout] = convertAudioSampleFormatToGStreamerFormat(self.format());
-    auto [destinationFormat, destinationLayout] = convertAudioSampleFormatToGStreamerFormat(format);
-    auto sourceOffset = frameOffset.value_or(0);
+    switch (format()) {
```
I wonder if this could be moved to the base class and have the avfoundation and gstreamer code use the same. It's getting pretty close to one another when converting format

philn replied:
About this specific case I'm not sure, the global approach is the same but the implementation differs quite a bit.

jyavenard commented on Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
```diff
@@ -188,67 +174,119 @@ int64_t PlatformRawAudioDataGStreamer::timestamp() const
+
+        auto copyElements = [numberOfChannels = numberOfChannels()]<typename T>(std::span<T> destination, const auto& source, size_t frames)
+        {
+            RELEASE_ASSERT(destination.size() >= frames * numberOfChannels);
```
same here

philn replied:
Yes I think this case is clear and the code can be shared.

jyavenard commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -840,6 +841,41 @@ GstAudioInfo* GstMappedAudioBuffer::info()
+    if (GST_AUDIO_INFO_LAYOUT(&m_buffer.info) == GST_AUDIO_LAYOUT_INTERLEAVED) {
+        auto planeSizeInBytes = (m_buffer.n_samples - offset) * GST_AUDIO_INFO_BPS(&m_buffer.info);
+        auto result = Vector<std::span<T>> { planeCount, [planeSizeInBytes](auto) {
+            return MallocSpan<T>::malloc(planeSizeInBytes).leakSpan();
```
isn't that actually leaking? the data will be copied into the Vector, but the underlying data isn't freed

philn replied:
Hmm I was naively supposing the storage would move to the vector, I'll check :)

jyavenard replied:
> Hmm I was naively supposing the storage would move to the vector, I'll check :)

The Vector constructor copy the data given in the span, no move.

calvaris commented on Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
```diff
@@ -188,67 +174,119 @@ int64_t PlatformRawAudioDataGStreamer::timestamp() const
-    GST_TRACE("Output caps: %" GST_PTR_FORMAT, outputCaps.get());
+    auto sourceFormat = audioData.format();
+    const auto& sourceSample = audioData.sample();
+    bool destinationIsInterleaved = isAudioSampleFormatInterleaved(format);
```
isDestinationInterleaved

calvaris reviewed (approved):
What jya said.

philn committed:
[GStreamer] imported/w3c/web-platform-tests/webcodecs/audio-data.any.html is a permanent failure
https://bugs.webkit.org/show_bug.cgi?id=284626

Reviewed by Xabier Rodriguez-Calvar.

The GStreamer AudioData::copyTo() is now spec compliant, handling the various layout conversion
cases (from/to planar/interleaved) and format changes.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-data.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-data.any.worker-expected.txt: Removed.
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp:
(WebCore::computeCopyElementCount):
* Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
(WebCore::GStreamerInternalAudioEncoder::encode):
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.cpp:
(WebCore::PlatformRawAudioData::create):
(WebCore::PlatformRawAudioDataGStreamer::numberOfFrames const):
(WebCore::PlatformRawAudioDataGStreamer::isInterleaved const):
(WebCore::PlatformRawAudioDataGStreamer::planesOfSamples):
(WebCore::layoutToString):
(WebCore::PlatformRawAudioData::copyTo):
(WebCore::getAudioConvertedForFormat): Deleted.
* Source/WebCore/platform/audio/gstreamer/PlatformRawAudioDataGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::GstMappedAudioBuffer::GstMappedAudioBuffer):
(WebCore::GstMappedAudioBuffer::samples const):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:

Canonical link: https://commits.webkit.org/290147@main
